### DISCUSSION TITLE
[HW Accel Support]: rpi5-64 (Raspberry Pi 5 64-bit OS) and Pineboard Coral Edge TPU

### DISCUSSION THREAD

**Discussion by Zaleo80** [blakeblackshear/frigate#11708] [Hardware Acceleration Support]
### Describe the problem you are having

When using any Coral M.2 Accelerator A+E Key Frigate Add-On does not start anymore.

Did try with bothe the Dual Edge Coral TPU Bundle from Pineboards and with a different HatDrive and a regular M.2 A+E-Key Google Coral Edge TPU.

As I am not sure where to reported I also reported it here: https://github.com/home-assistant/operating-system/issues/3279

Protected mode is disabled
![image](https://github.com/blakeblackshear/frigate/assets/128542757/9d18f36a-3579-40a8-b6f7-8e7ba1113c1a)

### Version

0.13.2-6476F8A

### Frigate config file

```yaml
mqtt:
  host: localhost
  user: mqtt-user
  password: ************

detectors:
  cpu1:
    type: cpu
    num_threads: 2
#  coral1:
#    type: edgetpu
#    device: pci:0

record:
  enabled: True
  retain:
    days: 7
    mode: motion
  events:
    retain:
      default: 30
      mode: motion

cameras:
  #  Voordeur:
  #    enabled: True
  #    ffmpeg:
  #      inputs:
  #        - path: rtsp://127.0.0.1:8554/T8210P002*****7
  #          roles:
  #            - detect
  Woonkamer:
    enabled: True
    ffmpeg:
      inputs:
        - path: rtsp://U2*****TQGu:84g******k@172.16.1.41/live0
          roles:
            - detect
```

### docker-compose file or Docker CLI command

```yaml
https://github.com/blakeblackshear/frigate/blob/dev/docker/rpi/Dockerfile
```

### Relevant log output

```shell
Process detector:coral1:
[2024-06-03 11:27:45] frigate.detectors.plugins.edgetpu_tfl ERROR   : No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
    delegate = Delegate(library, options)
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 119, in __init__
    raise ValueError(capture.message)
ValueError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/frigate/frigate/object_detection.py", line 102, in run_detector
    object_detector = LocalObjectDetector(detector_config=detector_config)
  File "/opt/frigate/frigate/object_detection.py", line 53, in __init__
    self.detect_api = create_detector(detector_config)
  File "/opt/frigate/frigate/detectors/__init__.py", line 18, in create_detector
    return api(detector_config)
  File "/opt/frigate/frigate/detectors/plugins/edgetpu_tfl.py", line 41, in __init__
    edge_tpu_delegate = load_delegate("libedgetpu.so.1.0", device_config)
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 162, in load_delegate
    raise ValueError('Failed to load delegate from {}\n{}'.format(
ValueError: Failed to load delegate from libedgetpu.so.1.0
```

### FFprobe output from your camera

```shell
Not related
```

### Operating system

HassOS

### Install method

HassOS Addon

### Network connection

Wired

### Camera make and model

Eufy Indoor Cam 2K Pan & Tilt

### Any other information that may be helpful

The /dev/apex_0 device does show up.

```
# dmesg | grep apex
[255773.884961] apex 0000:03:00.0: Couldn't reinit interrupts: -28
[255773.884996] apex 0000:03:00.0: Permission checking failed.
```

```
# lspci
03:00.0 Class 0880: 1ac1:089a
02:07.0 Class 0604: 1b21:1182
00:00.0 Class 0604: 14e4:2712
02:03.0 Class 0604: 1b21:1182
01:00.0 Class 0604: 1b21:1182
00:00.0 Class 0604: 14e4:2712
04:00.0 Class 0108: 1987:5013
01:00.0 Class 0200: 1de4:0001
```

**Comment by Zaleo80** [comment]
I tried to add a /etc/udev/rules.d/65-apex.rules file to set mode to 0660 for root:root.
The "Permission checking failed." error is now gone and somehow, we get a report about temperatures.

```
# dmesg | grep apex
[    1.898889] apex 0000:03:00.0: enabling device (0000 -> 0002)
[    1.899761] apex 0000:03:00.0: Couldn't initialize interrupts: -28
[    7.132396] apex 0000:03:00.0: Apex performance not throttled due to temperature
```
But Frigate is still not willing to find the TPU. 

Based on the remaining interrupts errors, I guess is that there is still something going on the MSI-X support, but I have never looked into Linux kernels with that kind of debth.

**Comment by srpablillo** [comment]
Interested on @Zaleo80 issue. Also trying to run Frigate with the Pineboard Dual Edge Coral TPU on a RP5 with HASOS

**Comment by danielszilagyi** [comment]
For what it's worth, I ran into the same issue. Hardware is RPi5 with MCUZone [MPDTPU](http://wiki.mcuzone.com/index.php?title=0025_MPDTPU(M.2_E)_EN)

 Solution was to switch to kernel8.img instead of the default kernel_2712.img and the apex driver stopped complaining on ```couldn't initialize interrupts``` immediately. Both corals detected and working after this change.
